### DISCUSSION TITLE
Header support for requests

### DISCUSSION THREAD

**Discussion by dagohan** [benweier/blizzard.js#127] [Ideas]
Hi 👋, awesome job on covering the battle-net API!

I checked if there is support to add headers to the requests, but couldn't find it and traced how requests are generated to here:
https://github.com/benweier/blizzard.js/blob/7e2aa4adcf0d23a7d648a24b4c7cabf924f99f01/src/core/index.ts#L80-L93

It is relying on default headers (also dynamically sets `Battlenet-Namespace` based on the resource) but other than that the user does not have any control over it.

I think it would be beneficial to allow users to pass optional headers to any resource. For example, some requests can potentially return large response bodies and the user might want to avoid this if the resource has not been updated between the previously observed state.  Blizzard API supports `If-Modified-Since` header for such purpose but it is not possible to utilize it behind `blizzard.js`'s abstraction.

**Comment by benweier** [comment]
Thanks @Dagohan this is a really good point you make. This was in the previous version but I didn't bring it over during the Typescript rewrite.

How do you feel about something like this

```ts
wow.characterProfile({ name: string, realm: string }, headers: { [key: string]: string })
```

Some API methods accept no arguments (mainly for "index" endpoints) so the first argument would become `undefined | null`

```ts
wow.achievements(undefined, headers: { [key: string]: string })
wow.achievements(null, headers: { [key: string]: string })
```

I think I had the foresight to also anticipate if an empty object is passed so this would also work for "index" endpoints

```ts
wow.achievements({}, headers: { [key: string]: string })
```

**Comment by benweier** [comment]
@Dagohan [`4.2.0`](https://github.com/benweier/blizzard.js/releases/tag/v4.2.0) has been published with support for `headers`

**Reply by dagohan** [reply]
wow, I didn't expect such quick reply & development. Thanks a lot @benweier, appreciated the support a lot :+1:

**Reply by benweier** [reply]
No problem, I'm still on holidays so I had the free time to focus on such a sensible feature request :)
